### PR TITLE
Ethernet bridging support

### PR DESCRIPTION
**Pull request by npitre**
This provides Ethernet bridging support to Zephyr, allowing two
or more Ethernet segment appear as if they were on the same network
to connected devices.

This is the basic support. Ethernet frames are copied to all
participating interfaces (but the source one) for now.

A shell to manipulate bridge configuration is also provided.

Future enhancements may include:

- MAC address caching to forward unicast pkts only to interfaces where 
  they were seen

- Special interface to act as the local access to the bridged network

- Spanning tree protocol

### CODE REVIEW

File: tests/net/bridge/prj.conf [hunk 1]
```diff
@@ -0,0 +1,14 @@
+CONFIG_NETWORKING=y
+CONFIG_NET_TEST=y
+CONFIG_NET_IPV4=n
+CONFIG_NET_IPV6=n
```
Comment (jukkar) [inline tests/net/bridge/prj.conf:4]: While this is fine if just bridging is enabled, it might be good to add an additional test to testcase.yaml that enables IPv4 and IPv6 so that we test that all works well in that case too.

**Review by jukkar — APPROVED** [review]
Nice :+1:, overall looks good, couple of minor comments.

**Review by rlubos — COMMENTED** [review]
Looks pretty good

**Comment by npitre** [comment]
Robert Lubos wrote:

> This description looks like a copypaste.

It is. Those functions are duplicating functionality found elsewhere but 
applied to bridges. So I used the same wording to document them.

**Comment by npitre** [comment]
Jukka Rissanen wrote:

> While this is fine if just bridging is enabled, it might be good to add an additional test to testcase.yaml that enables IPv4 and IPv6 so that we test that all works well in that case too.

Done.

**Comment by rlubos** [comment]
> This description looks like a copypaste.
> It is. Those functions are duplicating functionality found elsewhere but applied to bridges. So I used the same wording to document them.

Sure, but it's still using terms from net_capture, like `@typedef net_capture_cb_t`, it should be updated.

**Comment by npitre** [comment]
Robert Lubos wrote:

> Sure, but it's still using terms from net_capture, like ***@***.*** net_capture_cb_t`, it should be updated.

Ah! Indeed. I missed that twice now it seems. Done.

**Review by rlubos — APPROVED** [review]
LGTM, there are a few CI issues still to address though.

**Comment by npitre** [comment]
> LGTM, there are a few CI issues still to address though.

I know. Still trying to figure out a way to work around the obnoxiously 
inadequate CI web interface for those of us relying on accessibility tools.

File: include/net/net_pkt.h [hunk 2]
```diff
@@ -181,7 +181,6 @@ struct net_pkt {
 					* segment.
 					*/
 #endif
-
```
Comment (jukkar) [inline include/net/net_pkt.h]: Unrelated change, should be removed

File: tests/net/bridge/prj_ip.conf [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+CONFIG_NET_L2_ETHERNET=y
+CONFIG_NET_ETHERNET_BRIDGE=y
+CONFIG_NET_LOG=y
+CONFIG_ENTROPY_GENERATOR=y
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_NET_PKT_TX_COUNT=10
+CONFIG_NET_PKT_RX_COUNT=10
+CONFIG_NET_BUF_RX_COUNT=10
+CONFIG_NET_BUF_TX_COUNT=10
+CONFIG_ZTEST=y
```
Comment (jukkar) [inline tests/net/bridge/prj_ip.conf:14]: I propose not to introduce any new prj.conf file for this but just add a config to testcase.yaml file, see my other comment there.

File: tests/net/bridge/testcase.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+tests:
+  net.eth_bridge:
+    min_ram: 16
+    tags: net bridge
+    depends_on: netif
+  net.eth_bridge.ip:
+    min_ram: 16
+    tags: net bridge
+    depends_on: netif
+    extra_args: CONF_FILE="prj_ip.conf"
```
Comment (jukkar) [inline tests/net/bridge/testcase.yaml:10]: Please change this to
```
common:
  min_ram: 16
  tags: net bridge
  depends_on: netif
tests:
  net.eth_bridge:
    extra_configs:
      - CONFIG_NET_IPV4=n
      - CONFIG_NET_IPV6=n
  net.eth_bridge.ip:
    extra_configs:
      - CONFIG_NET_IPV4=y
      - CONFIG_NET_IPV6=y
```
The extra config in `net.eth_bridge` is duplicate (as it is already in prj.conf file) but something is needed in that section as the section cannot be left empty. If done this way, there is no need to have extra prj config file and we need to maintain only one prj file in the tests.

**Comment by jukkar** [comment]
There is this warning
```
Run if [[ -s "output.txt" ]]; then
Error: subsys/net/l2/ethernet/bridge_shell.c:14:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:27:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:68:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:101:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:176:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
Error: Process completed with exit code 1.
```

In the header file documentation you are missing `endcond`
```
/home/runner/work/zephyr/zephyr/include/net/ethernet_bridge.h:31: Conditional section with label 'INTERNAL_HIDDEN' does not have a corresponding \endcond command within this file.
```

File: subsys/net/l2/ethernet/bridge.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+enum net_verdict net_eth_bridge_input(struct ethernet_context *ctx,
+				      struct net_pkt *pkt);
+
+static inline bool net_eth_iface_is_bridged(struct ethernet_context *ctx)
+{
+#if defined(CONFIG_NET_ETHERNET_BRIDGE)
+	return ctx->bridge.instance != NULL;
+#else
+	return FALSE;
```
Comment (jukkar) [inline subsys/net/l2/ethernet/bridge.h:18]: s/FALSE/false/

**Comment by nashif** [comment]
```


In file included from /workdir/zephyr/subsys/net/l2/ethernet/ethernet.c:33:
--
  | /workdir/zephyr/subsys/net/l2/ethernet/bridge.h: In function 'net_eth_iface_is_bridged':
  | /workdir/zephyr/subsys/net/l2/ethernet/bridge.h:18:9: error: 'FALSE' undeclared (first use in this function)
  | 18 \|  return FALSE;
  | \|         ^~~~~


```

**Comment by nashif** [comment]
```
Error: subsys/net/l2/ethernet/bridge_shell.c:14:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:27:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:68:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:101:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
subsys/net/l2/ethernet/bridge_shell.c:176:WARNING: Violation to rule 5.7 (Tag name should be unique) tag: shell
Error: Process completed with exit code 1.
```

**Comment by nashif** [comment]
```
Warning, treated as error:
/home/runner/work/zephyr/zephyr/include/net/ethernet_bridge.h:31: Conditional section with label 'INTERNAL_HIDDEN' does not have a corresponding \endcond command within this file.
FAILED: CMakeFiles/html 
```

**Comment by npitre** [comment]
Thanks forall your help.

There are still issues with buildkit: 
https://buildkite.com/zephyr/zephyr/builds/35972

Of course twister is happy here.

**Comment by nashif** [comment]
```
FAILED: zephyr/zephyr_prebuilt.elf zephyr/zephyr_prebuilt.map 
: && ccache /opt/toolchains/zephyr-sdk-0.12.4/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc   zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj -o zephyr/zephyr_prebuilt.elf  -fuse-ld=bfd  -Wl,-T  zephyr/linker_zephyr_prebuilt.cmd  -Wl,-Map=/workdir/zephyr/twister-out/ip_k66f/tests/net/bridge/net.eth_bridge/zephyr/zephyr_prebuilt.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/boards/arm/ip_k66f/libboards__arm__ip_k66f.a  zephyr/subsys/net/libsubsys__net.a  zephyr/subsys/net/l2/ethernet/libsubsys__net__l2__ethernet.a  zephyr/subsys/net/ip/libsubsys__net__ip.a  zephyr/subsys/random/libsubsys__random.a  zephyr/subsys/testsuite/ztest/libsubsys__testsuite__ztest.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/spi/libdrivers__spi.a  zephyr/drivers/entropy/libdrivers__entropy.a  modules/nxp/lib..__..__var__lib__buildkite-agent__zephyr-module-cache__modules__hal__nxp.a  modules/segger/libmodules__segger.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -L"/opt/toolchains/zephyr-sdk-0.12.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/thumb/v7e-m/nofp"  -L/workdir/zephyr/twister-out/ip_k66f/tests/net/bridge/net.eth_bridge/zephyr  -lgcc  zephyr/arch/common/libisr_tables.a  -mcpu=cortex-m4  -mthumb  -mabi=aapcs  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -no-pie  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn  -Wl,--fatal-warnings && cd /workdir/zephyr/twister-out/ip_k66f/tests/net/bridge/net.eth_bridge/zephyr && /usr/local/bin/cmake -E echo
/opt/toolchains/zephyr-sdk-0.12.4/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: warning: orphan section `._eth_bridge.static.test_bridge' from `app/libapp.a(main.c.obj)' being placed in section `._eth_bridge.static.test_bridge'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

guess this one: ```warning: orphan section `._eth_bridge.static.test_bridge' from `app/libapp.a(main.c.obj)' being placed in section `._eth_bridge.static.test_bridge'```

**Comment by npitre** [comment]
Anas Nashif wrote:

> guess this one: ```warning: orphan section `._eth_bridge.static.test_bridge' from `app/libapp.a(main.c.obj)' being placed in section `._eth_bridge.static.test_bridge'```

What is the platform/test combination producing this?

**Comment by npitre** [comment]
Figured it out. CI is now happy, and so do I.
